crates/red_knot/src/ast_ids.rs

```diff
@@ -0,0 +1,322 @@
+use std::fmt::Formatter;
+use std::marker::PhantomData;
+
+use rustc_hash::FxHashMap;
+
+use ruff_index::{Idx, IndexVec};
+use ruff_python_ast::visitor::preorder;
+use ruff_python_ast::visitor::preorder::{PreorderVisitor, TraversalSignal};
+use ruff_python_ast::{
+    AnyNodeRef, AstNode, ModModule, NodeKind, Parameter, Stmt, StmtAnnAssign, StmtAssign,
+    StmtAugAssign, StmtClassDef, StmtFunctionDef, StmtImport, StmtImportFrom, StmtTypeAlias,
+    TypeParam, TypeParamParamSpec, TypeParamTypeVar, TypeParamTypeVarTuple,
+};
+use ruff_text_size::{Ranged, TextRange};
+
+#[ruff_index::newtype_index]
+pub struct AstId;
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub struct FileAstId<N: HasAstId> {
```

Comment:
Why do we need this wrapper around an `AstId`?

Reply:
This is related to the PhantomData. See https://github.com/astral-sh/ruff/pull/10849#discussion_r1562158940